feat: rework x/sched callbacks and remove cosmos/evm fork

pitasi opened the pull request:
This PR end goal is to remove our cosmos/evm fork, which was needed because our callbacks were not emitting ethereum logs properly.

For achieving that:
- a new transaction is injected at every block (see x/sched's PrepareProposal and ProcessProposal). This new transaction is a MsgEthereumTx, signed using a random private key (the signature won't be checked). The transaction will include a new extension option.
- the injected transaction, is processed differently than normal MsgEthereumTxs by the ante handler, that recognizes the new extension option. **This transaction skips classic security checks such as signature and nonce**. Normal users shouldn't be able to send this kind of transactions themselves.
- When executed, the injected transaction triggers a special call in the x/sched precompile that will execute all pending callbacks. This is another change from the previous architecture: other modules (e.g. x/async) won't be able to directly trigger a callback, but only to schedule them. The actual callback will be executed during the next block.

I added a new integration test that uses `forge` to build, deploy and interact with a contract that uses x/async and waits for two callbacks.

pitasi committed:
tests: callbacks integration test

pitasi committed:
fix(wardend): remove our fork of cosmos/evm

pitasi committed:
fix(wardend): move Proposal handlers to new abciutil package

Instead of using vemanager, which should only contain vote extensions.

pitasi committed:
feat(x/sched): add new ExtensionOptionsCallbacks

This allow to make a separate ante handler for the new special kind of
transaction that will be built by x/sched.

pitasi committed:
feat(x/sched): inject an eth tx invoking executeCallbacks()

This tx will trigger x/sched precompile, and will be present at every
block.

The precompile will be able to fetch all the pending callbacks and do
its work.

The ante handler and ProcessProposal ensures that malicious users or
validators can't add this type of transaction to the blocks manually,
but only the injected transaction will work.

pitasi committed:
feat(x/sched): implement executeCallbacks() precompile fn

pitasi committed:
chore: update CHANGELOG

backsapc commented on precompiles/sched/sched.go:
```diff
@@ -99,6 +102,22 @@ func (p *Precompile) Run(evm *vm.EVM, contract *vm.Contract, readonly bool) (bz
+	case ExecuteCallbacksMethod:
+		bz, err = p.ExecuteCallbacksMethod(ctx, evm, stateDB, method, args)
+
+	// TODO: remove following comments, it looks like it works anyway now
```
So you want to remove?

pitasi replied:
ah yes i forgot, removed now

thanks :D

mn13 commented on tests/framework/exec/forge.go:
Can't we just use bindings for creating and calling contracts? Bindings & go-ethereum allows to do most of the things that forge gives (see example [here](https://github.com/warden-protocol/wardenprotocol/blob/d804a23bdda6e40449f9adb5436d92e44bd0ec26/tests/cases/json_precompile_user.go#L31))

pitasi replied:
i refactored some utils for working with ethereum tests and used the bindings instead of forge

pitasi committed:
feat(tests): refactor ethereum tests

I'm trying to reduce the number of calls and make the tests as readable
as possible.
